flang/lib/Lower/ConvertExpr.cpp

```diff
@@ -1646,28 +1675,64 @@ class ScalarExprLowering {
       }
     }
 
-    // Handle case where caller must pass result
-    auto resRef = [&]() -> llvm::Optional<fir::ExtendedValue> {
+    // Handle cases where caller must allocate the result or a fir.box for it.
+    if (caller.mustMapInterfaceSymbols())
+      TODO(loc, "function result that requires mapping interface symbols");
+    auto idxTy = builder.getIndexType();
+    auto lowerSpecExpr = [&](const auto &expr) -> mlir::Value {
+      return builder.createConvert(
+          loc, idxTy, fir::getBase(converter.genExprValue(expr, stmtCtx)));
+    };
+    llvm::SmallVector<mlir::Value> resultLengths;
+    auto allocatedResult = [&]() -> llvm::Optional<fir::ExtendedValue> {
+      llvm::SmallVector<mlir::Value> extents;
+      llvm::SmallVector<mlir::Value> lengths;
+      if (!caller.callerAllocateResult())
+        return {};
+      auto type = caller.getResultStorageType();
+      if (type.isa<fir::SequenceType>())
+        caller.walkResultExtents([&](const Fortran::lower::SomeExpr &e) {
+          extents.emplace_back(lowerSpecExpr(e));
+        });
+      caller.walkResultLengths([&](const Fortran::lower::SomeExpr &e) {
+        lengths.emplace_back(lowerSpecExpr(e));
+      });
```

Comment:
Is this walking of lengths and extents ever done one without the other? They seem very similar. Can they be folding together?

Reply:
In the pointer/allocatable result case, the lengths are always walked, but not the extents.